### PR TITLE
fix: Report Service

### PR DESCRIPTION
**Pull request by DerekCrannaford**
Good UserWhere From AllCalls

Report UserWhere being called by array, string and 'empty'
Then breaking anyway due to incorrect quote escapes
Fixes for versatile calls & forced SQL escaping
Tested Support:
/report/index?module=
/report/reportAjaxCategorytoType?id=_
/report/reportAjaxListModules

refs: clients assert reports dont work

### CODE REVIEW

**Review by strongishllama — CHANGES_REQUESTED** [review]
Looks good, just remove that commented var_dump on line 188. I'd be keen for @adam-buckley to approve this as well though before merging it. Just due to the importance of reports.

**Review by strongishllama — CHANGES_REQUESTED** [review]
Looks good! Just add the comments and type hinting as we discussed, cheers.

**Comment by DerekCrannaford** [comment]
Added commit to include patches on RunReport vs ExeReport for membership rules to match those of getUserWhere
Added SQLI unit test for getUserWhere : php cmfive.php tests unit report

File: system/modules/report/models/ReportService.php [hunk 1]
```diff
@@ -27,7 +27,16 @@ public function getReportMembers($id)
...
     public function getReportMember($id, $uid)
     {
-        return $this->getObject("ReportMember", array("report_id" => $id, "user_id" => $uid, "is_deleted" => 0));
+        $conferred = [];
+        $conferred[] = $this->getObject("ReportMember", ["report_id" => $id, "user_id" => $uid, "is_deleted" => 0]);
+        $groups = AuthService::getInstance($this->w)->getGroups();
+
+        foreach ($groups ?? [] as $group) {
+            if (AuthService::getInstance($this->w)->getUser($uid)->inGroup($group)) {
+                $conferred[] = $this->getObject("ReportMember", ["report_id" => $id, "user_id" => $group->id, "is_deleted" => 0]);
```
Comment (adam-buckley) [inline system/modules/report/models/ReportService.php:36]: Thinking about this again, there's still a chance that the member return here is null. before the return below we should be calling array_filter over this conferred variable which will strip out null values

File: system/modules/report/models/ReportService.php [hunk 2]
```diff
@@ -144,25 +153,27 @@ public function getReportsbyUserWhere($id, $where)
...
         // Clause for admin user
-        if ($this->w->Auth->user()->hasRole("report_admin")) {
+        if (AuthService::getInstance($this->w)->user()->hasRole("report_admin")) {
             return $this->getReports();
         }
 
         // need to get reports for me and my groups
         // me
-        $myid = [$id];
+        $myid[] = $this->_db->quote($id);
```
Comment (adam-buckley) [inline system/modules/report/models/ReportService.php:162]: This is not the correct way to instantiate an array, even though this works it should be ```$myid = [$this->_db->quote($id)];```

File: system/modules/report/models/ReportService.php [hunk 3]
```diff
@@ -176,11 +187,45 @@ public function getReportsbyUserWhere($id, $where)
...
-            Report::$_db_table . " as r on m.report_id = r.id where m.user_id in (" . $this->_db->quote($theid) . ") " . (!empty($where) ? $this->_db->quote($where) : '') .
-            " and r.is_deleted = 0 order by r.is_approved desc,r.title")->fetch_all();
+            Report::$_db_table . " as r on m.report_id = r.id " .
+            " where m.user_id in (" . $theid . ") " . $filter .
+            " and r.is_deleted = 0 and m.is_deleted = 0 " .
+            " order by r.is_approved desc,r.title")->fetch_all();
         return $this->fillObjects("Report", $rows);
     }
 
+     public function unitaryWhereToAndClause($where) {
```
Comment (adam-buckley) [inline system/modules/report/models/ReportService.php:197]: Not PSR2 compliant

**Review by adam-buckley — CHANGES_REQUESTED** [review]

File: system/modules/report/models/ReportService.php (hunk 1 above)
Comment (DerekCrannaford) [reply system/modules/report/models/ReportService.php:36]: +1

File: system/modules/report/models/ReportService.php (hunk 3 above)
Comment (DerekCrannaford) [reply system/modules/report/models/ReportService.php:197]: +1

File: system/modules/report/models/ReportService.php (hunk 2 above)
Comment (DerekCrannaford) [reply system/modules/report/models/ReportService.php:162]: OK

**Comment by DerekCrannaford** [comment]
Any suggested extensions/additions to unit test?

**Review by adam-buckley — APPROVED** [review]

**Pull request merged**